recipes/join-tables.md

```diff
@@ -0,0 +1,385 @@
+
+When retrieving multiple users, we can capitalize on the fact that we know that
+for our own context, we want to return a _paginated_ set of results, and thus
+return a pagination adapter. The appropriate one for us is
+`Zend\Paginator\Adapter\DbSelect`, which expects the `Select` instance itself,
+plus a database adapter and result set prototype &mdash; which we can retrieve
+from the table gateway itself.
+
+So, we now have a `TableGateway`. How do we get it properly instantiated and
+configured, though? We need to make sure it is getting a `HydratingResultSet` so
+that we get back our `UserEntity` instances, and we need to make sure it's
+getting the correct table name and database adapter.
+
+We can do that by creating a factory. A `Zend\Db\TableGateway\TableGateway`
+instance expects up to five arguments:
+
+- The table name.
+- The zend-db adapter to use.
+- An array of [features](https://docs.zendframework.com/zend-db/table-gateway/#tablegateway-features),
+  if any are needed.
+- A result set prototype instance.
+- A `Sql` instance to use as a prototype.
+
+zf-apigility generally provides the first four arguments only, and supplies a
+`null` value for the third. We'll do similarly in the factory we create.
+
+```php
+// in modules/Users/src/V1/Rest/Users/UsersTableGatewayFactory.php:
+namespace Users\V1\Rest\Users;
+
+use Psr\Container\ContainerInterface;
+use Zend\Db\ResultSet\HydratingResultSet;
+use Zend\Hydrator\ArraySerializable;
+
+class UsersTableGatewayFactory
+{
+    public function __invoke(ContainerInterface $container)
+    {
+        return new UsersTableGateway(
+            'users',
+            $container->get('user_locations'),
+            null,
+            $this->getResultSetPrototype($container)
+        );
+    }
+
+    private function getResultSetPrototype(ContainerInterface $container)
+    {
+        $hydrators = $container->get('HydratorManager');
+        $hydrator = $hydrators->get(ArraySerializable::class);
+        return new HydratingResultSet($hydrator, new UsersEntity());
+    }
+}
+```
+
+We know all the values we need up front; it's just a matter of pulling what we
+need from the container and/or creating instances.
+
+In `modules/Users/config/module.config.php`, update the `service_manager`
+configuration as follows:
+
+```php
+'service_manager' => [
+    'factories' => [
+        Users\V1\Rest\Users\UsersTableGateway::class => Users\V1\Rest\Users\UsersTableGatewayFactory::class,
+    ],
+],
+```
+
+We now have a table gateway capable of producing what we need. We now need to
+tell the resource to use these new methods.
+
+### Updating the resource
+
+As noted earlier, DB-Connected services have a "virtual" resource. zf-apigility
+creates a `ZF\Apigility\DbConnectedResource` instance that performs the various
+operations. We need to somehow alter this to use our new table gateway, and also
+use the new methods it presents. How do we do that?
```

Comment:
I’m not a fan of this sentence as it doesn’t really describe what we want to happen.

We don’t want to alter the virtual resource, we want to replace it with our own resource class that adds our new additonal functionality to what we get with the DbConnectedResource.

No idea how to phrase that though, and the current sentence can be lived with.